File: go-backend-interview-prep/04_interfaces_and_generics/08_generics_constraints/main.py
# Python equivalent of Generics Constraints (compare with main.go)
# Go constraints ~ Python TypeVar bounds or Protocol constraints

from typing import TypeVar

# --- comparable: Python has __eq__ on everything, so no special constraint ---
T = TypeVar("T")


def index(s: list[T], target: T) -> int:
    """Python has list.index(), but this matches Go's generic Index."""
    for i, v in enumerate(s):
        if v == target:
            return i
    return -1


# --- Number constraint: TypeVar with union ---
# Python TypeVar with constraints restricts to specific types
Num = TypeVar("Num", int, float)


def sum_all(s: list[Num]) -> Num:
    total = type(s[0])(0) if s else 0
    for v in s:
        total += v
    return total


# --- Ordered: Python supports < > on int, float, str natively ---
Ord = TypeVar("Ord", int, float, str)


def max_val(a: Ord, b: Ord) -> Ord:
    return a if a > b else b


def min_val(a: Ord, b: Ord) -> Ord:
    return a if a < b else b


# --- Named types (like Go's type UserID int) ---
class UserID(int):
    """Named type based on int (like Go's ~int with type alias)."""
    pass


def main():
    # --- comparable ---
    print("--- comparable (index) ---")
    nums = [10, 20, 30, 40]
    print("index of 30:", index(nums, 30))
    print("index of 99:", index(nums, 99))

    words = ["go", "python", "rust"]
    print("index of python:", index(words, "python"))

    # --- Number constraint ---
    print("\n--- Number constraint (sum) ---")
    ints = [1, 2, 3, 4, 5]
    print("sum ints:", sum_all(ints))

    floats = [1.1, 2.2, 3.3]
    print("sum floats:", sum_all(floats))

    # --- Named type based on int ---
    # Python: subclass of int works with int operations
    ids = [UserID(1), UserID(2), UserID(3)]
    print("sum UserIDs:", sum(ids))

    # --- Ordered ---
    print("\n--- Ordered (max/min) ---")
    print("max(3, 7):", max_val(3, 7))
    print("min(3, 7):", min_val(3, 7))
    print("max(a, z):", max_val("a", "z"))

    # Key difference:
    # Go constraints are enforced at compile time -- wrong type = compile error.
    # Python TypeVar bounds are hints only -- runtime does not enforce them.


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/04_interfaces_and_generics/08_generics_constraints/main.go
package main

import (
	"cmp"
	"fmt"
)

// --- comparable: allows == and != ---
func Index[T comparable](s []T, target T) int {
	for i, v := range s {
		if v == target {
			return i
		}
	}
	return -1
}

// --- Custom constraint: Number ---
type Number interface {
	~int | ~int64 | ~float64
}

func Sum[T Number](s []T) T {
	var total T
	for _, v := range s {
		total += v
	}
	return total
}

// --- ~ matches underlying types ---
type UserID int // underlying type is int

// --- Ordered constraint (uses cmp.Ordered from Go 1.21+) ---
func Max[T cmp.Ordered](a, b T) T {
	if a > b {
		return a
	}
	return b
}

func Min[T cmp.Ordered](a, b T) T {
	if a < b {
		return a
	}
	return b
}

// --- Constraint with method requirement ---
type Stringer interface {
	String() string
}

func PrintAll[T Stringer](items []T) {
	for _, item := range items {
		fmt.Println(" ", item.String())
	}
}

type Tag struct{ Value string }

func (t Tag) String() string { return "#" + t.Value }

func main() {
	// --- comparable ---
	fmt.Println("--- comparable (Index) ---")
	nums := []int{10, 20, 30, 40}
	fmt.Println("index of 30:", Index(nums, 30))
	fmt.Println("index of 99:", Index(nums, 99))

	words := []string{"go", "python", "rust"}
	fmt.Println("index of python:", Index(words, "python"))

	// --- Custom Number constraint ---
	fmt.Println("\n--- Number constraint (Sum) ---")
	ints := []int{1, 2, 3, 4, 5}
	fmt.Println("sum ints:", Sum(ints))

	floats := []float64{1.1, 2.2, 3.3}
	fmt.Println("sum floats:", Sum(floats))

	// --- ~ matches named types ---
	ids := []UserID{1, 2, 3}
	fmt.Println("sum UserIDs:", Sum(ids)) // works because ~int matches UserID

	// --- cmp.Ordered (allows < > comparisons) ---
	fmt.Println("\n--- Ordered (Max/Min) ---")
	fmt.Println("max(3, 7):", Max(3, 7))
	fmt.Println("min(3, 7):", Min(3, 7))
	fmt.Println("max(a, z):", Max("a", "z"))

	// --- Constraint with method ---
	fmt.Println("\n--- Stringer constraint ---")
	tags := []Tag{{Value: "go"}, {Value: "generics"}, {Value: "interview"}}
	PrintAll(tags)
}
